Feature/task 2

VlastimilM opened the pull request.

VlastimilM committed:
KA-131 - Create project structure

Add project Contracts and ListItemRepository

VlastimilM committed:
KA-131 - Join ListItemController with ListItemRepository

VlastimilM committed:
KA-131 - Fix project names and structure, references

VlastimilM committed:
KA-131 - Add constructor taking repository to ListItemController

VlastimilM committed:
Merge branch 'feature/task-1' into feature/task-2

VlastimilM committed:
KA-131 - Fix ListItem after merge conflict

VlastimilM committed:
KA-131 - Adjust tests with mocked repository using NSubstitute

VlastimilM committed:
KA-131 - Add ContainerBootstrapper class

VlastimilM committed:
KA-131 - Add UnityResolver, UnityResolverConfig

VlastimilM committed:
KA-131 - Clean solution using ReSharper

VlastimilM committed:
KA-131 - Change status code in PostAsync of ListItemsController

VlastimilM committed:
KA-131 - Remove redundant PostAsync test

VlastimilM committed:
KA-131 - Add Validation to PostAsync method

Commit:
KA-131 - Create ResponseConfig

It ensures a lower camel case response

Commit:
KA-131 - Rename WebApiConfig to RouteConfig

LindaH32 committed:
KA-131 - Add RouteConfig

LindaH32 committed:
KA-131 - Change return type of RegisterRepositoryTypes

LindaH32 committed:
KA-131 - Add guid check to PostAsync, adjust tests

LindaH32 committed:
KA-131 - Remove unused imports

LindaH32 committed:
KA-131 - Refactor tests

LindaH32 committed:
KA-131 - Change ListItem model to mutable

VlastimilM committed:
KA-131 - Fix ListItem initializations

VlastimilM committed:
KA-131 - Create extension method for EqualConstraint

The extension method UsingListItemComparer compares two list items.
Refactor multiple methods into lambdas.

JiriLojda commented on KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs:
```diff
@@ -1,44 +1,61 @@
+            }
 
-            return Created("api/v1/items/?id=300...", await Task.FromResult(SampleItems[0]));
+            return Created("api/v1/items/?id=300...", await Task.FromResult(_repository.Post(item)));
```
The route you return as a location produces an error. It would be much better if it would show me the item I just created using this Post method.
ApiController has an [Url](https://msdn.microsoft.com/en-us/library/system.web.http.apicontroller.url(v=vs.108).aspx?cs-save-lang=1&cs-lang=csharp#code-snippet-1) property which will help you generate URLs. You may also want to inject some class which will create it for you. See "One might..." bullet in the task description.

VlastimilM committed:
KA-131 - Adjust Dispose method in UnityResolver

VlastimilM committed:
KA-131 - Refactor Dispose method in UnityResolver

JiriLojda commented on KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs:
```diff
@@ -1,44 +1,61 @@
+        {
+            if (item == null)
+            {
+                ModelState.AddModelError(string.Empty, "Item is null");
```
You are using static property Empty on class String, not type string. So, although it works this way, it would look better if you put here String explicitly.

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+            new ListItem{ Id = new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), Text = "updated"}
+        };
+
+        public IEnumerable<ListItem> Get() 
```
Methods in a repository (working with a database) should be async since you do not want to block until database response come.

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    internal class ListItemRepository : IListItemRepository
+    {
+        private static readonly List<ListItem> SampleItems = new List<ListItem>
```
List looks more like you want to add or delete items in it. Since you do not intend to add or delete items in SampleItems, it would look better to just create Array for it.

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+        private static readonly List<ListItem> SampleItems = new List<ListItem>
+        {
+            new ListItem { Id= new Guid("98DBDE18-639E-49A6-8E51-603CEB2AE92D"), Text= "text" },
```
So let's find two missing spaces. :)

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+        {
+            new ListItem { Id= new Guid("98DBDE18-639E-49A6-8E51-603CEB2AE92D"), Text= "text" },
+            new ListItem { Id = new Guid("1C353E0A-5481-4C31-BD2E-47E1BAF84DBE"), Text = "giraffe" },
+            new ListItem{ Id = new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), Text = "updated"}
```
And one missing space here...

JiriLojda commented on KenticoListApp/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs:
```diff
@@ -4,163 +4,163 @@
 
         [Test]
         public void PostAsync_WithValidArguments_ReturnsCorrectItemAndStatusCode()
         {
```
You should also test returned location. You would find out, that location you are returning now is not correct. :D

JiriLojda commented on KenticoListApp/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs:
```diff
@@ -4,163 +4,163 @@
         }
 
         [Test]
         public void GetAsync_WithoutParams_ReturnsTestItems()
```
Forgotten status code check?

JiriLojda commented on KenticoListApp/TodoList.Api.Tests/Helpers/EqualConstraintExtensions.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace TodoList.Api.Tests.Helpers
+{
+    public static class EqualConstraintExtensions
```
This can be internal...

JiriLojda commented on KenticoListApp/TodoList.Api.Tests/Helpers/ListItemComparer.cs:
```diff
@@ -1,5 +1,5 @@
+using TodoList.Contracts.Models;
 
 namespace TodoList.Api.Tests.Helpers
 {
```
This class can also be internal.

JiriLojda commented on KenticoListApp/TodoList.Api/Resolvers/UnityResolver.cs:
```diff
@@ -0,0 +1,63 @@
+            }
+            catch (ResolutionFailedException)
+            {
+                return new List<object>();
```
Since you are returning just some empty IEnumerable, it would be more readable to return Enumerable.Empty as a sign that it is just some empty Enumerable.

JiriLojda commented on KenticoListApp/TodoList.Api/Resolvers/UnityResolver.cs:
```diff
@@ -0,0 +1,63 @@
+            return new UnityResolver(child);
+        }
+
+        public void Dispose()
```
Surely you already know how to fix this.

JiriLojda commented on KenticoListApp/TodoList.Api/App_Start/ResponseConfig.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace TodoList.Api
+{
+    public static class ResponseConfig
```
Can be internal.

JiriLojda commented on KenticoListApp/TodoList.Api/App_Start/RouteConfig.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace TodoList.Api
+{
+    public static class RouteConfig
```
Can be internal.

JiriLojda commented on KenticoListApp/TodoList.Api/App_Start/UnityResolverConfig.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace TodoList.Api
+{
+    public static class UnityResolverConfig
```
Also can be internal.

JiriLojda commented on KenticoListApp/TodoList.Api/App_Start/WebApiConfig.cs:
```diff
@@ -2,13 +2,10 @@
 
 namespace TodoList.Api
 {
-    public static class WebApiConfig
```
Surely you already know about this.

JiriLojda commented on KenticoListApp/TodoList.Api/App_Start/UnityResolverConfig.cs:
```diff
@@ -0,0 +1,17 @@
+        public static void Register(HttpConfiguration config)
+        {
+            var container = new UnityContainer()
+                .RegisterRepositoryTypes();
```
Imagine, you want to register classes in another assembly for DI in this architecture. You have to look how other registrations work. Then you find out you have to create another extension on IUnityContainer, which will register your types.
What I would appreciate more (when creating new assemblies for your project) is to create a class implementing some defined interface (e.g. IBootstrapper). That interface would tell me what should I implement. Then use some generic method to tell UnityResolverConfig, that I want to register types also from my class.

JiriLojda commented on KenticoListApp/TodoList.Api/Controllers/ListItemsController.cs:
```diff
@@ -1,44 +1,61 @@
+            _repository = repository;
+        }
+        
         public async Task<IHttpActionResult> PostAsync(ListItem item)
```
In my opinion, methods here should be in order Get, Post, Put, Delete. But it is just a matter of taste...

JiriLojda reviewed (changes requested):
Good job, there are only a few things, that can be even better. :)

There is only one bigger problem. You have committed many garbage file. You have to double check what you are about to commit. With such files committed into the draft solution, you could break it.
![garbagefilesintask2review](https://user-images.githubusercontent.com/25480503/27135063-dfbefde4-5117-11e7-86fe-3d98705afe6d.png)

Aside from two bigger things and these garbage files, it is all just minor details.

Firstly, a common structure for assemblies is to have sources in Sources folder and tests in Tests folder.

Secondly, please rename your solution to match names of your projects (TodoList). You can also rename project names to match solution name, but it would require much more work. :D

When moving assemblies into folders, you will have to update paths in the solution file. Close Visual Studio and use Notepad++ or you favorite text editor for it, solution file is just a text file with some information about its projects. You will also have to update some paths in project files, those having references on other projects. Use Notepad++ also for this as well.

JiriLojda commented on KenticoListApp/TodoList.Contracts/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,39 @@
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+
+[assembly: AssemblyTitle("Contracts")]
```
This should match project name, but it does not, does it?

JiriLojda commented on KenticoListApp/TodoList.Contracts/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,39 @@
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Kentico software")]
+[assembly: AssemblyProduct("Contracts")]
```
Same for this.

JiriLojda commented on KenticoListApp/TodoList.Repositories/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,39 @@
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+
+[assembly: AssemblyTitle("ListItemRepository")]
```
And this.

JiriLojda commented on KenticoListApp/TodoList.Repositories/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,39 @@
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Kentico software")]
+[assembly: AssemblyProduct("ListItemRepository")]
```
And this.

VlastimilM committed:
KA-131 - Change config files from public to internal

VlastimilM committed:
KA-131 - Refactor SampleItems in ListItemRepository into array

VlastimilM committed:
KA-131 - Change helper classes in tests to internal

VlastimilM committed:
KA-131 - Remove unnecessary files in Models

VlastimilM committed:
KA-131 - Remove garbage file from Models

VlastimilM committed:
KA-131 - Add status code check to GetAsync test

VlastimilM committed:
KA-131 - Refactor container registrations

Create IBootstrapper

VlastimilM committed:
KA-131 - Create ListItemsGenerator [In progress]

LindaH32 committed:
KA-131 - Fix renaming of WebApiConfig (RouteConfig)

LindaH32 committed:
KA-131 - Rename WebApiConfig to RouteConfig

LindaH32 committed:
KA-131 - Fix renaming in TodoList.Api.csproj

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+            new ListItem { Id = new Guid("D69E065C-99B1-4A73-B00C-AD05F071861F"), Text = "updated" }
+        };
+
+        public IEnumerable<ListItem> Get() 
```
There is no need for overload here. Using GetAll would be much more expressive.

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+        public ListItem Delete(Guid id) 
+            => SampleItems[1];
+
+        public ListItem Put(ListItem item) 
```
Again using Update instead of Put would be much more expressive. You don't have to use name according to  Http verbs here. In fact, you shouldn't use it since this repository can be used for some other Presentation layer.

JiriLojda commented on KenticoListApp/TodoList.Repositories/ListItemRepository.cs:
```diff
@@ -0,0 +1,32 @@
+        public ListItem Put(ListItem item) 
+            => SampleItems[2];
+
+        public ListItem Post(ListItem item) 
```
Similarly to the previous method, this can be Create instead of Post.

VlastimilM committed:
KA-131 - Add ApiBootstrapper, Refactor ListItemUrlGenerator

VlastimilM committed:
KA-131 - Create IListItemUrlGenerator

Adjust PostAsync_WithValidArguments test

VlastimilM committed:
KA-131 - Make ApiBootstrapper register ListItemsUrlGenerator

Tests now test returned location

VlastimilM committed:
KA-131 - Change ListItemRepository methods to async

VlastimilM committed:
KA-131 - Rename methods in IListItemRepository

VlastimilM committed:
KA-131 - Correct AssemblyInfo files

VlastimilM committed:
KA-131 - Rename solution to TodoList

VlastimilM committed:
KA-131 - Add correct Repositories.csproj file

VlastimilM committed:
KA-131 - Move projects to folders Sources and Tests

LindaH32 committed:
KA-131 - Change methods order in ListItemRepository

LindaH32 committed:
KA-131 - Make ResponseConfig internal

LindaH32 committed:
KA-131 - Change members of internal classes to public

VlastimilM committed:
KA-131 - Fix PostAsync so it returns correct location

VlastimilM committed:
KA-131 - Change ListItemUrlGenerator to internal

VlastimilM committed:
KA-131 - Rename ListItemController tests

VlastimilM committed:
KA-131 - Refactor ListItemsController

VlastimilM committed:
KA-131 - Make ListItem initializations in test consistent

PetrSvirak commented on TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs:
```diff
@@ -18,10 +18,16 @@ public ListItemsController(IListItemRepository repository, IListItemUrlGenerator
         }
 
         public async Task<IHttpActionResult> GetAsync()
-            => Ok(await _repository.GetAllAsync());
```
Why did you do this? I found no mention of something like this in the review nor your commit comment justifying it; also it would strongly suggest against this step. Same applies for the `Put` and `Delete`…

PetrSvirak commented on TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs:
```diff
@@ -18,10 +18,16 @@ public ListItemsController(IListItemRepository repository, IListItemUrlGenerator
-            => Ok(await _repository.GetAsync(id));
+        {
+            var item = await _repository.GetAsync(id);
+            return Ok(item);
```
I do miss `id` validation here. You will want to check for the item of given identifier existence in `task 3`, however, you can surely tell even that now that there is one ID that will (or at least should) never exist in your DB (`Guid.Empty`).

PetrSvirak commented on TodoList/Sources/TodoList.Api/Services/ListItemUrlGenerator.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        private readonly UrlHelper _helper;
+
+        public ListItemUrlGenerator(HttpRequestMessage message)
```
You can even ask for `UrlHelper` helper here, knowing that constructor with most parameters will be used, it will work nicely for you. Just consider this.

PetrSvirak commented on TodoList/Sources/TodoList.Api/ApiBootstrapper.cs:
```diff
@@ -0,0 +1,17 @@
+        public IUnityContainer Register(IUnityContainer container)
+            => container
+                .RegisterType<HttpRequestMessage>(
+                    new InjectionFactory(_ => (HttpRequestMessage) HttpContext.Current.Items["MS_HttpRequestMessage"]))
```
Move the `HttpRequestMessage` obtainment into a named method. Moreover, you should feel some small urge to justify yourself for such code and put a nice explanatory comment above the named method :stuck_out_tongue:

PetrSvirak commented on TodoList/Sources/TodoList.Api/ApiBootstrapper.cs:
```diff
@@ -0,0 +1,17 @@
+            => container
+                .RegisterType<HttpRequestMessage>(
+                    new InjectionFactory(_ => (HttpRequestMessage) HttpContext.Current.Items["MS_HttpRequestMessage"]))
+                .RegisterType<IListItemUrlGenerator, ListItemUrlGenerator>();
```
In both these registrations, scoped lifetime manager is missing.

PetrSvirak commented on TodoList/Sources/TodoList.Contracts/Interfaces/IListItemUrlGenerator.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using TodoList.Contracts.Models;
+
+namespace TodoList.Contracts.Interfaces
```
I'm not completely against this interface being contracts, however, it will always be tightly coupled with .Api, so unless you decided to create a separate assembly for services related to the web (.Services.Web or .Services.Api, ...), I strongly suggest you move this interface right next to its implementation.

PetrSvirak commented on TodoList/Sources/TodoList.Repositories/RepositoryBootstrapper.cs:
```diff
@@ -0,0 +1,11 @@
+    public class RepositoryBootstrapper : IBootstrapper
+    {
+        public IUnityContainer Register(IUnityContainer container) 
+            => container.RegisterType<IListItemRepository, ListItemRepository>(new TransientLifetimeManager());
```
This is not the correct lifetime manager. You need to use the one that is scoped (based on our verbal agreement and reasoning). Read more on [Unity lifetime manager](https://msdn.microsoft.com/en-us/library/ff660872(v=pandp.20).aspx), please.

PetrSvirak commented on TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs:
```diff
@@ -1,166 +1,178 @@
+        }
+
+        [Test]
+        public void Delete_ReturnsCorrectResponse()
```
The tested method is not named `Delete` anymore…

PetrSvirak commented on TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs:
```diff
@@ -1,166 +1,178 @@
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+            Assert.That(error.ModelState.Keys, Is.EqualTo(expectedKeys));
+            Assert.That(actualLocation, Is.EqualTo(null));
```
This actually looks like forget temporary fix. There is much better, more descriptive, way to test for null – simply (assert that <something>) `Is.Null`) – fix all occurences, of course.

PetrSvirak commented on TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs:
```diff
@@ -1,166 +1,178 @@
+            var postedListItem = new ListItem { Id = Guid.Empty, Text = "newText" };
+            var expectedLocation = new Uri($"api/v1/ListItems/{_guidOfFirstItem}", UriKind.Relative);
+            _repository.CreateAsync(postedListItem).Returns(expectedListItem);
+            _urlGenerator.GenerateUrl(expectedListItem).Returns($"api/v1/ListItems/{expectedListItem.Id}");
```
You should actually test whether correct item got passed to the method. Compose the substitute in a way it reads id from the argument passed to the method (you can leverage the [`callInfo` that NSubstitute provides](http://nsubstitute.github.io/help/return-from-function/))

PetrSvirak commented on TodoList/Tests/TodoList.Api.Tests/Helpers/ListItemComparer.cs:
```diff
@@ -1,18 +1,18 @@
+
+namespace TodoList.Api.Tests.Helpers
+{
+    internal class ListItemComparer : IEqualityComparer<ListItem>
```
Consider going a step further and -- as this comparer will be used nowhere but in `EqualConstraintExtensions` -- moving it into the `EqualConstraintExtensions.cs` as a private class.

PetrSvirak reviewed (changes requested):
It is getting much better quickly. Nice work, guys :-)

#### UnityResolver
There are two unresolved issue with `UnityResolver`
* This way, your `Dispose` is not sufficient. What about inheriting classes? 
* The class has more extensive modifier than it needs, fix this as well, please.

#### Interfaces namespace
The `.Contract.Interfaces` namespace will only clutter you referenced types while it grows (in time).

For example, `IBootstrapper` interfaces will likely be used in other parts (and context) of your application than `IListItemRepository`. 

Moreover, they represent different concepts. The later is implemented in one assembly while the former in all of them. 

In general, it is not bad practice to couple interfaces by assemblies in which they are implemented (or more precisely, couple related things to assemblies and then create separate assemblies for these assemblies' contracts). 

Please, fix this by moving the interfaces into separate, well-named namespaces within the .Contracts assembly.

#### Renegade commit
The d988c02d535dfa4cbed9abb84298da4ee9effe06 has really poor description – it is not cleared what work had been done and why. Omit similar commit(s) in the future, please.

VlastimilM committed:
KA-131 - Change multiple methods in ListItemsController to lambdas

Rename repository to listItemsRepository

LindaH32 committed:
KA-131 - Make UnityResolver internal and sealed

LindaH32 committed:
KA-131 - Move interfaces into separate namespaces

LindaH32 committed:
KA-131 - Validate id in GetAsync, add related test

LindaH32 committed:
KA-131 - Add lifetimeManager to ApiBoostrapper registrations

VlastimilM committed:
KA-131 - Test _urlGenerator in PostAsync receives listItem

Rename Delete test to DeleteAsync
Make IListItemUrlGenerator ready for moving to TodoList.Api.Services

VlastimilM committed:
KA-131 - Extract Message aquisition in ApiBootstrapper

LindaH32 committed:
KA-131 - Move item comparer into EqualConstraintExtensions

LindaH32 committed:
KA-131 - Add correct lifeTime managers to registrations

LindaH32 committed:
KA-131 - Remove unnecessary white lines

PetrSvirak commented on TodoList/Sources/TodoList.Api/ApiBootstrapper.cs:
```diff
@@ -2,16 +2,21 @@
+            .RegisterType<HttpRequestMessage>(
+                new HierarchicalLifetimeManager(), 
+                new InjectionFactory(_ => GetCurrentContextRequestMessage()))
+            .RegisterType<IListItemUrlGenerator, ListItemUrlGenerator>(new HierarchicalLifetimeManager());
```
I kinda liked the previous indentation, however, this one is up to you.

PetrSvirak commented on TodoList/Sources/TodoList.Api/ApiBootstrapper.cs:
```diff
@@ -2,16 +2,21 @@
 using TodoList.Api.Services;
-using TodoList.Contracts.Interfaces;
+using TodoList.Contracts.Api;
+using IListItemUrlGenerator = TodoList.Api.Services.IListItemUrlGenerator;
```
Why do you need this?

PetrSvirak commented on TodoList/Sources/TodoList.Api/Controllers/ListItemsController.cs:
```diff
@@ -1,78 +1,81 @@
+        {
+            ValidatePostItem(item);
+
+            if (!ModelState.IsValid)
```
You might want to be a bit more consistent between methods. Here, I think the `if` is more related to the `ValidatePostItem` than to item creation, very similar to the case of `ValidateItemId` in `Get`. Thus the free line should rather be present between the `if`'s end and item creation.

PetrSvirak commented on TodoList/Sources/TodoList.Contracts/Api/IBootstrapper.cs:
```diff
@@ -1,6 +1,6 @@
 ﻿using Microsoft.Practices.Unity;
 
-namespace TodoList.Contracts.Interfaces
+namespace TodoList.Contracts.Api
```
`Bootstrapper` is in no way connected with (Web) Api while this namespace suggests such meaning.  Move it to a `.Bootstrap` namespace, please. 

The point there is that this interface is used in many (right now two, true) assemblies and they using statements should not express any particular connection to any particular user-facing facility. For example, if there was a WPF application, it would also need to call (all) bootstraps during its startup, yet it would have nothing in common with (web) API…

PetrSvirak commented on TodoList/Sources/TodoList.Contracts/Interfaces/IListItemUrlGenerator.cs:
```diff
@@ -1,9 +0,0 @@
-﻿using TodoList.Contracts.Models;
```
And here you lost your history. @VlastimilM, correct me if I'm wrong, but you it was your commit that doomed the file history, right? So, please, do fix this up so the whole interface lifetime can be seen. None of the commits currently available in the history of the `IListiItemUrl` in the `.Api` (70851b37d9e939e788ba8265b4a4ed847eac2f8e, 0871c8d509db380c0f4375ad087deeda2408eaf3) explains nor justifies its existence.

PetrSvirak commented on TodoList/Tests/TodoList.Api.Tests/Controllers/ListItemsControllerTest.cs:
```diff
@@ -1,178 +1,187 @@
+            var actionResult = _controller.GetAsync(Guid.Empty).Result;
+            var responseMessage = actionResult.ExecuteAsync(CancellationToken.None).Result;
+
+            Assert.That(responseMessage.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
```
The error message (or at least the message's key) should be asserted here as well.

PetrSvirak reviewed (changes requested):
There are 3 or 4 nitpicking details that remain. But also 1 incomplete test and 1 task that won't be that easy (chasing lost memories (~ history) is never easy, but I'm sure @VlastimilM will fight it well :crossed_fingers:). Otherwise, the task is complete. We are almost there and you did really nice job overall, thanks.

VlastimilM committed:
KA-131 - Remove unnecessary using in ApiBootstrapper

Remove unnecessary attribute in IListItemUrlGenerator

VlastimilM committed:
KA-131 - Move ApiBootstrapper into Bootstrap namespace

VlastimilM committed:
KA-131 - Add error message keys assert to GetAsync test

VlastimilM committed:
KA-131 - Move IBootstrapper into Bootstrap namespace

VlastimilM committed:
KA-131 - Move IListItemUrlGenerator to TodoList.Api.Services

LindaH32 committed:
KA-131 - Rename PutAsync test method for consistency

LindaH32 committed:
KA-131 - Change method order in IListItemRepository
